src/backend/joanie/core/api.py

```diff
@@ -127,6 +137,124 @@ def perform_create(self, serializer):
+
+        try:
+            product = Product.objects.get(uid=product_id)
+        except Product.DoesNotExist:
+            return Response(f"Product {product_id} does not exist.", status=404)
+
+        # If product is not free, we have to create a payment.
+        # To create one, a billing address is mandatory
+        if product.price.amount > 0 and billing_address is None:
+            return Response({"billing_address": "This field is required."}, status=400)
+
+        # - Validate data then create an order
+        serializer = self.get_serializer(data=request.data)
+        try:
+            serializer.is_valid(raise_exception=True)
+            self.perform_create(serializer)
+        except (DRFValidationError, IntegrityError):
+            return Response(
+                (
+                    f"Cannot create order related to the product {product_id} "
+                    f"and course {course_code}"
+                ),
+                status=400,
+            )
+
+        # Once order has been created, if product is not free, create a payment
+        if product.price.amount > 0:
+            order = serializer.instance
+            payment_backend = get_payment_backend()
+
+            if credit_card_id:
+                try:
+                    credit_card = CreditCard.objects.get(
+                        owner=order.owner, uid=credit_card_id
+                    )
+                    payment_info = payment_backend.create_one_click_payment(
+                        request=request,
+                        order=order,
+                        billing_address=billing_address,
+                        credit_card_token=credit_card.token,
+                    )
+                    return Response(
+                        {**serializer.data, "payment_info": payment_info}, status=201
+                    )
+                except (CreditCard.DoesNotExist, NotImplementedError):
+                    pass
+            else:
+                payment_info = payment_backend.create_payment(
+                    request=request, order=order, billing_address=billing_address
+                )
+
+            # Return the fresh new order with payment_info
+            return Response(
+                {**serializer.data, "payment_info": payment_info}, status=201
+            )
+
+        # Else return the fresh new order
+        return Response(serializer.data, status=201)
+
+    @action(detail=True, methods=["POST"])
+    def abort(self, request, uid=None):  # pylint: disable=no-self-use
+        """Abort a pending order and the related payment if there is one."""
+        payment_id = request.data.get("payment_id")
+
+        try:
+            user = models.User.objects.get(username=request.user.username)
+            order = models.Order.objects.get(uid=uid, owner=user)
+        except ObjectDoesNotExist:
+            return Response("Cannot found the order ressource.", status=404)
```

Comment:
Maybe give more details?
```suggestion
            return Response("No order found with id \"{uid}\" owned by {request.user.username}.", status=404)
```